docs/src/main/tut/datatypes/nel.md

```diff
@@ -0,0 +1,182 @@
+  None
+} else {
+  Some(xs.sum / xs.length.toDouble)
+}
+```
+
+That works and is safe, but this only masks the problem of *accepting
+invalid input*.  By using `Option`, we extend the `average` function
+with the logic to handle empty lists.  Additionally, all callers have
+to handle the `Option` cases, maybe over and over again.  While better
+than failing with an exception, this is far from perfect.
+
+Instead what we would like to express is that `average` does not make
+sense at all for an empty list.  Luckily, cats defines the
+`NonEmptyList`.  As the name says, this represents a list that
+*cannot*, by construction, be empty.  So given a `NonEmptyList[A]` you
+can be sure that there is at least one `A` in there.
+
+Let's see how that impacts your `average` method:
+
+```tut:silent
+import cats.data.NonEmptyList
+def average(xs: NonEmptyList[Int]): Double = {
+  xs.reduceLeft(_+_) / xs.length.toDouble
+}
+```
+
+With that, `average` is free of any "domain invariant validation" and
+instead can focus on the actual logic of computing the average of the
+list.  This ties in nicely with the recommendation of shifting your
+validation to the very borders of your program, where the input enters
+your system.
+
+## Structure of a NonEmptyList
+
+`NonEmptyList` is defined as follows:
+
+```scala
+final case class NonEmptyList[+A](head: A, tail: List[A]) {
+  // Implementation elided
+}
+```
+
+The `head` of the `NonEmptyList` will be _non-empty_. Meanwhile, the
+`tail` can have zero or more elements contained in a `List`.
+
+## Defined for all its elements
+
+An important trait of `NonEmptyList` is the totality. For `List` specifically,
+both `head` and `tail` are partial: they are only well-defined if it
+has at least one element.
+
+`NonEmptyList` on the other hand, _guarantees_ you that operations like
+`head` and `tail` are defined, because constructing an empty
+`NonEmptyList` is simply not possible!
+
+## Constructing a NonEmptyList
+
+To construct a `NonEmptyList` you have different possibilities.
+
+### Using `one`
+
+If you want to construct a collection with only one argument, use
+`NonEmptyList.one`:
+
+```tut
+NonEmptyList.one(42)
+```
+
+### Using `of`
+
+The `NonEmptyList.of` method on the companion of `NonEmptyList` as the signature:
+
+```scala
+def of[A](head: A, tail: A*): NonEmptyList[A]
+```
+
+It accepts an argument list with at least one `A` followed by a
+*varargs* argument for the `tail`.  Call it like this:
+
+```tut
+NonEmptyList.of(1)
+NonEmptyList.of(1, 2)
+NonEmptyList.of(1, 2, 3, 4)
+```
+
+There also is `ofInitLast` which takes a normal `List[A]` for the
+prefix and a last element:
+
+```tut
+NonEmptyList.ofInitLast(List(), 4)
+NonEmptyList.ofInitLast(List(1,2,3), 4)
+```
+
+### Using `fromList`
+
+There is also `NonEmptyList.fromList` which returns an
+`Option[NonEmptyList[A]]`:
+
+```tut
+NonEmptyList.fromList(List())
+NonEmptyList.fromList(List(1,2,3))
```

Comment:
Minor: if you have imported `cats.implicits._` there's also some syntax for a `toNel` method added to `List`. It's quite a bit more concise than `NonEmptyList.fromList`, so it may be worth giving an example.